File: seminar/src/main/java/com/server/dosopt/seminar/controller/PostControllerV2.java
package com.server.dosopt.seminar.controller;

import com.server.dosopt.seminar.dto.request.post.PostCreateRequest;
import com.server.dosopt.seminar.dto.request.post.PostUpdateRequest;
import com.server.dosopt.seminar.dto.response.post.PostGetResponse;
import com.server.dosopt.seminar.service.PostService;
import com.server.dosopt.seminar.service.PostServiceV2;
import java.net.URI;
import java.security.Principal;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

@Slf4j
@RestController
@RequestMapping("/api/v2/post")
@RequiredArgsConstructor
public class PostControllerV2 {

    private final PostServiceV2 postService;

    private static final String CUSTOM_AUTH_ID = "X-Auth-id";  // 상수(static final)로 관리

    @PostMapping
    public ResponseEntity<Void> createPost(@RequestPart MultipartFile image,
                                            PostCreateRequest request,
                                            Principal principal) {
        log.info("컨트롤러 요청 성공");
        Long memberId = Long.valueOf(principal.getName());
        URI location = URI.create("/api/post" + postService.createV2(request, image, memberId));
        return ResponseEntity.created(location).build();
    }

    @GetMapping("/{postId}")
    public ResponseEntity<PostGetResponse> getPostById(@PathVariable Long postId) {
        return ResponseEntity.ok(postService.getById(postId));
    }

    @GetMapping
    public ResponseEntity<List<PostGetResponse>> getPosts(@RequestHeader(CUSTOM_AUTH_ID) Long memberId){
        return ResponseEntity.ok(postService.getPosts(memberId));
    }

    @PatchMapping("/{postId}")
    public ResponseEntity<Void> updatePost(@PathVariable Long postId,
                                           @RequestBody PostUpdateRequest request) {
        postService.update(request, postId);
        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{postId}")
    public ResponseEntity<Void> deletePost(@PathVariable Long postId) {
        postService.deleteByIdV2(postId);
        return ResponseEntity.noContent().build();
    }
}

File: kotlin-seminar/src/main/kotlin/sopt/org/kotlinseminar/controller/PostControllerV2.kt
package sopt.org.kotlinseminar.controller

import jakarta.validation.Valid
import lombok.RequiredArgsConstructor
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import sopt.org.kotlinseminar.dto.request.post.PostCreateRequest
import sopt.org.kotlinseminar.dto.request.post.PostUpdateRequest
import sopt.org.kotlinseminar.dto.response.post.PostGetResponse
import sopt.org.kotlinseminar.enums.Constant.CUSTOM_AUTH_ID
import sopt.org.kotlinseminar.service.PostService
import sopt.org.kotlinseminar.service.PostServiceV2
import java.net.URI

@RestController
@RequestMapping("/api/v2/post")
@RequiredArgsConstructor
class PostControllerV2(
        private val postService: PostServiceV2
) {

    @PostMapping
    fun createPost(@RequestHeader(CUSTOM_AUTH_ID) memberId: Long,
                    @RequestPart image: MultipartFile,
                    @Valid @RequestBody request: PostCreateRequest): ResponseEntity<Void>? {
        val postId = postService.createV2(request, image, memberId)
        val location: URI = URI.create("/api/v2/post" + postId)
        return ResponseEntity.created(location).build()
    }

    @DeleteMapping("/{postId}")
    fun deletePost(@PathVariable postId: Long): ResponseEntity<Void> {
        postService.deleteByIdV2(postId)
        return ResponseEntity.noContent().build()
    }
}